cpp/src/transform/jit/kernel.cu

```diff
@@ -66,59 +100,81 @@ struct scalar {
   using type                     = typename Accessor::type;
   static constexpr int32_t index = Accessor::index;
 
-  static __device__ decltype(auto) element(cudf::mutable_column_device_view_core const* views,
+  static __device__ decltype(auto) element(cudf::mutable_column_device_view_core const* outputs,
+                                           void* user_data,
                                            cudf::size_type row)
   {
-    return Accessor::element(views, 0);
+    return Accessor::element(outputs, user_data, 0);
   }
 
-  static __device__ decltype(auto) element(cudf::column_device_view_core const* views,
+  static __device__ decltype(auto) element(cudf::column_device_view_core const* inputs,
+                                           void* user_data,
                                            cudf::size_type row)
   {
-    return Accessor::element(views, 0);
+    return Accessor::element(inputs, user_data, 0);
   }
 
-  static __device__ void assign(cudf::mutable_column_device_view_core const* views,
+  static __device__ void assign(cudf::mutable_column_device_view_core const* outputs,
                                 cudf::size_type row,
                                 type value)
   {
-    return Accessor::assign(views, 0, value);
+    return Accessor::assign(outputs, 0, value);
   }
 };
 
 template <typename Out, typename... In>
-CUDF_KERNEL void kernel(cudf::mutable_column_device_view_core const* output,
-                        cudf::column_device_view_core const* inputs)
+CUDF_KERNEL void kernel(cudf::mutable_column_device_view_core const* outputs,
+                        cudf::column_device_view_core const* inputs,
+                        void* user_data)
 {
   // cannot use global_thread_id utility due to a JIT build issue by including
   // the `cudf/detail/utilities/cuda.cuh` header
   auto const block_size          = static_cast<thread_index_type>(blockDim.x);
   thread_index_type const start  = threadIdx.x + blockIdx.x * block_size;
   thread_index_type const stride = block_size * gridDim.x;
-  thread_index_type const size   = output->size();
+  thread_index_type const size   = outputs[0].size();
 
   for (auto i = start; i < size; i += stride) {
-    GENERIC_TRANSFORM_OP(&Out::element(output, i), In::element(inputs, i)...);
+    GENERIC_TRANSFORM_OP(&Out::element(outputs, user_data, i),
+                         In::element(inputs, user_data, i)...);
```

Comment:
Why not just this?
```suggestion
    GENERIC_TRANSFORM_OP(user_data, &Out::element(outputs, i),
                         In::element(inputs, i)...);
```

Reply:
But we know that from the transform API parameter.

Reply:
Yes, that is what I was expecting. I think I'd rather have a separate kernel with user-data and one without.

Reply:
Why is that?
We already have 3 kernels for transforms; the fixed-width, fixed-point, and span kernel. 
By having separate kernels we'd have 6 kernels that do similar work with no advantage in debuggability, maintainability, or performance.
Considering we might even decide to support structs and lists in the future, that would make 10 kernels.
I personally would have even preferred to make them all into a single kernel.

Reply:
It's fine to try to reduce duplication but I just don't agree with this approach to do it I suppose.
```
    GENERIC_TRANSFORM_OP(&Out::element(outputs, user_data, i),
                         In::element(inputs, user_data, i)...);
```

Reply:
We could just require all UDF kernels to accept a user-data first parameter and we just pass through what we get. Then we don't make the parameter optional but default it to nullptr.

Reply:
> We could just require all UDF kernels to accept a user-data first parameter and we just pass through what we get. Then we don't make the parameter optional but default it to nullptr.

That would break compatibility with Python/Numba, the ABI currently doesn't take a user_data, and Numba doesn't support `void *` types.
It'd also mean updating the existing tests to match the signature.

Reply:
> It's fine to try to reduce duplication but I just don't agree with this approach to do it I suppose.

Yes, I know it seems gnarly but it makes the code cleaner.
An alternative would be to pack them into a tuple and tag the accessors using indices to select their arguments.

```
struct user_data_accessor{
static constexpr int arg = 2;

// element() will now use its arg index to unpack from the tuple

};
```

```
std::tuple<cudf::mutable_column_view*, cudf::column_view *, void *> args;
```

```
GENERIC_TRANSFORM_OP(&Out::element(args, i), In::element(args, i)...);
```

Reply:
Sorry, I don't think the `tuple` is right either. 
I'd thought the output part would be a single element/column of a `mutable_column_view` and there would be one or more input arguments of type `column_view` (core one). This makes it look like the `Out` part contains a `column_view` and user data.

Reply:
Yes, the point of the accessors is to disambiguate the variadic types and unpack each function argument from the kernel's arguments; it's different from column_view's accessors.

The solution of duplicating it works as well and doesn't break NUMBA compatibility, should I go with that?